File: Scripts/Scala/DistanceJoin/src/main/scala/TaskCoreMapper.scala
package edu.ucr.dblab.djoin

import org.slf4j.{LoggerFactory, Logger}
import scala.collection.JavaConverters._
import scala.io.Source
import java.io.FileWriter

object TaskCoreMapper{
  implicit val logger: Logger = LoggerFactory.getLogger("myLogger")

  case class Task(taskId: Int, host: String, launch: Long, finish: Long, duration: Int){
    def getLaunch: Int = launch.toInt
    def getFinish: Int = finish.toInt

    override def toString(): String = s"$host\t$taskId\t$launch\t$finish\t$duration\n"
  }

  case class Matrix(cores: Int, times: Int){
    val T = (0 to cores).map(n => Array.fill(times){0}).toVector
    val empty = '-'
    val busy  = '*'

    def set(core: Int, time: Int, taskId: Int): Unit = T(core)(time) = taskId

    def get(core: Int, time: Int): Int = T(core)(time)

    def update(core: Int, u: Array[Int], start: Int, duration: Int): Unit = {
      u.copyToArray(T(core), start, duration)
    }

    def getAvailableCore(time: Int): Int = {
      val candidates = (0 to cores).map(core => T(core)(time))
      val core = candidates.indexOf(0)
      if(core >= 0){
        core
      } else {
        getAvailableCore(time + 1)
      }
    }

    override def toString(): String = {
      (0 to cores).map{ core =>
        s"$core : " +: T(core).map{x => if(x == 0) empty else busy}.mkString("")
      }.mkString("\n")
    }
  }

  def main(args: Array[String]): Unit = {
    implicit val params = new TaskCoreMapperConf(args)
    val filename = params.filename()
    val cores = params.cores()
    val maxt = params.maxt()

    logger.info("Reading file...")
    val bufferedSource = Source.fromFile(filename)
    val tasks_prime = bufferedSource.getLines.drop(1).map{ line =>
      val arr = line.split("\t")
      val taskId   = arr(3).toInt
      val launch   = arr(4).toLong
      val finish   = arr(5).toLong
      val duration = arr(6).toInt
      val host     = arr(9)

      Task(taskId, host, launch, finish, duration)
    }.toVector

    logger.info("Setting start and end...")
    val start = tasks_prime.map(_.launch).min
    val tasks = tasks_prime.map{ t =>
      t.copy(launch = t.launch - start, finish = t.finish - start)
    }//.filter(_.host == "mr-01")
    val end = tasks.map(_.finish).max

    val times = end.toInt + 1
    val matrix = Matrix(cores, times)
    println(times)

    val f = new FileWriter("/tmp/tcmapper.tsv")
    f.write("coreId\thost\ttaskId\tlaunchTime\tfinishTime\tduration\n")
    for(task <- tasks){
      val core = matrix.getAvailableCore(task.getLaunch)
      f.write(s"${core}\t${task.toString}")
      val interval = Array.fill(task.duration){task.taskId}
      matrix.update(core, interval, task.getLaunch, task.duration)
    }
    f.close

    logger.info("Done!")
    bufferedSource.close
  }
}

File: Scripts/Python/TaskCoreMapper.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Jul  7 20:32:25 2020

@author: and
"""
from dataclasses import dataclass

@dataclass(frozen=True)
class Task:
    host: str
    taskId: int
    launch: int
    finish: int
    duration: int
    
    def toString(self):
        return("{}\t{}\t{}\t{}\t{}\n".format(host, taskId, launch, finish, duration))

id = 17
filename = "/home/and/Research/Meetings/next/figures/pflock{}.tsv".format(id)
cores = 8

def getAvailableCore(T, time, taskId):
    try:
        core = [T[i][time] for i in range(cores)].index(0)
        return(core)
    except ValueError:
        #print("Not available core at {} for task {}".format(time, taskId))
        return(getAvailableCore(T, time + 1, taskId))

tasks_prime = []
with open(filename) as f:
    next(f)
    for line in f:
        arr = line.split("\t")
        taskId = int(arr[3])
        launch = int(arr[4])
        finish = int(arr[5])
        duration = int(arr[6])
        host = arr[9]
        task = Task(host, taskId, launch, finish, duration)
        tasks_prime.append(task)
start = min(tasks_prime, key = lambda task: task.launch).launch    
tasks = []
for t in tasks_prime:
    task = Task(t.host, t.taskId, t.launch - start, t.finish - start, t.duration)
    tasks.append(task)
length = max(tasks, key = lambda  task: task.finish).finish

f = open("/home/and/Research/Meetings/next/figures/mapper{}.tsv".format(id), "w")
f.write("coreId\thost\ttaskId\tlaunchTime\tfinishTime\tduration\n")

hosts = ["mr-01","mr-02","mr-03","mr-04","mr-05","mr-06","mr-07","mr-08","mr-09","mr-10","mr-11","mr-12"]

for host in hosts:
    print(host)
    T = [[0 for i in range(length)] for j in range(cores)]
    sample = [task for task in tasks if task.host == host]
    sample.sort(key = lambda task: task.launch)
    for t in sample:
        core = getAvailableCore(T, t.launch, taskId = t.taskId)
        line = "{}\t{}\t{}\t{}\t{}\t{}\n".format(core, 
                                                 t.host, 
                                                 t.taskId, 
                                                 t.launch+start, 
                                                 t.finish+start, 
                                                 t.duration)
        f.write(line)
        interval = [x for x in range(t.launch, t.finish)]
        for time in interval:
            T[core][time] = t.taskId
f.close()